### PR TITLE
Tahoe-LAFS

### PR DESCRIPTION
**Pull request by JosephKiranBabu**
Tahoe-LAFS is a Free and Open decentralized cloud storage system. 
This may be used for fault-tolerant backups of files stored on a network of FreedomBoxes.

### CODE REVIEW

**Comment by JosephKiranBabu** [comment]
Closed and reopened pull request to trigger Travis CI build.

**Comment by jvalleroy** [comment]
I tried a quick test (in vagrant) and ran into this error, seems to be a permissions problem:
```
[2017-04-08 20:03:34,964] django.request ERROR    Internal Server Error: /plinth/apps/tahoe-lafs/
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/django/core/handlers/exception.py", line 39, in inner
    response = get_response(request)
  File "/usr/lib/python3/dist-packages/django/core/handlers/base.py", line 249, in _legacy_get_response
    response = self._get_response(request)
  File "/usr/lib/python3/dist-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/lib/python3/dist-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/lib/python3/dist-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/Plinth-0.13.1-py3.5.egg/plinth/modules/tahoe/views.py", line 63, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/django/views/generic/base.py", line 88, in dispatch
    return handler(request, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/django/views/generic/edit.py", line 174, in get
    return self.render_to_response(self.get_context_data())
  File "/usr/local/lib/python3.5/dist-packages/Plinth-0.13.1-py3.5.egg/plinth/modules/tahoe/views.py", line 68, in get_context_data
    context['introducers'] = tahoe.get_introducers()
  File "/usr/local/lib/python3.5/dist-packages/Plinth-0.13.1-py3.5.egg/plinth/modules/tahoe/__init__.py", line 218, in get_introducers
    with open(introducers_file, 'r') as intro_conf:
PermissionError: [Errno 13] Permission denied: '/var/lib/tahoe-lafs/storage_node/private/introducers.yaml'
```

**Comment by jvalleroy** [comment]
I'm still getting the permissions issue (with a fresh vagrant box). Would it be better to always read from /var/lib/tahoe-lafs/private as tahoe-lafs user?

**Comment by lispyclouds** [comment]
@jvalleroy Fixed the issues. Please check.

File: .gitignore [hunk 1]
```diff
@@ -23,3 +23,4 @@ plinth/tests/coverage/report/
 .vagrant/
 .idea/
 .DS_Store
+.eggs
```
Comment (jvalleroy) [inline .gitignore]: Let's have a newline at the end of each file, for consistent style.

File: LICENSES [hunk 2]
```diff
@@ -61,3 +61,7 @@ otherwise.
 - static/themes/default/icons/transmission.png :: [[https://transmissionbt.com/][GPL]]
 - static/themes/default/icons/ttrss.png :: [[https://tt-rss.org/gitlab/fox/tt-rss][GPL]]
 - static/themes/default/icons/ejabberd.png :: [[https://www.ejabberd.im/][GPL-2]]
+- static/themes/default/icons/xmpp.png :: [[https://www.ejabberd.im/][GPL3]]
+- static/themes/default/icons/openvpn.png :: [[https://github.com/thekishanraval/Logos][GPLv3+]]
+- static/themes/default/icons/syncthing.png :: [[https://github.com/syncthing/syncthing/][Mozilla Public License Version 2.0]]
+- static/themes/default/icons/tahoe.png :: [[https://en.wikipedia.org/w/index.php?curid=31565848][Creative Commons Attribution 3.0 Unported License]]
```
Comment (jvalleroy) [inline LICENSES]: Please keep these sorted alphabetically. Also xmpp.png can be removed.

File: actions/tahoe-lafs [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    subcommand = arguments.subcommand.replace('-', '_')
+    subcommand_method = globals()['subcommand_' + subcommand]
+    subcommand_method(arguments)
+
+
+if __name__ == '__main__':
+    main()
+
+
+class DomainNameNotSetupException(Exception):
```
Comment (jvalleroy) [inline actions/tahoe-lafs]: You can import this from plinth.exception.

File: actions/tahoe-lafs [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    setup.add_argument('--domain-name',
+                       help='The domain name to be used by Tahoe-LAFS')
+    subparsers.add_parser('create-introducer',
+                          help='Create and start the introducer node')
+    subparsers.add_parser('create-storage-node',
+                          help='Create and start the storage node')
+    subparsers.add_parser('autostart',
+                          help="Automatically start all introducers and "
+                               "storage nodes on system startup")
+    intro_parser_add = subparsers.add_parser('add-introducer',
```
Comment (jvalleroy) [inline actions/tahoe-lafs]: If you move 'add-introducer' down to the next line, the help text can take up a wider space, and be easier to read.

File: actions/tahoe-lafs [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+                                                       "be a tuple of "
+                                                       "(pet_name, furl)")
+    intro_parser_remove = subparsers.add_parser('remove-introducer',
+                                                help="Rename the introducer "
+                                                     "entry in the "
+                                                     "introducers.yaml file "
+                                                     "specified by the param")
+    intro_parser_remove.add_argument(
+        '--pet-name', help='The domain name that will be used by '
+                           'Matrix Synapse')
```
Comment (jvalleroy) [inline actions/tahoe-lafs]: -> Tahoe-LAFS

File: actions/tahoe-lafs [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+                               "on this FreedomBox.")
+    subparsers.add_parser('get-local-introducer',
+                          help="Return the name and furl of the introducer "
+                               "created on this FreedomBox")
+
+    return parser.parse_args()
+
+
+def subcommand_setup(arguments):
+    """Actions to be performed before installing Tahoe-LAFS."""
```
Comment (jvalleroy) [inline actions/tahoe-lafs]: Should be "after installing".

File: actions/tahoe-lafs [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+    # Create tahoe group if needed.
+    try:
+        grp.getgrnam('tahoe-lafs')
+    except KeyError:
+        subprocess.run(['addgroup', 'tahoe-lafs'], check=True)
+
+    # Create tahoe user is needed.
```
Comment (jvalleroy) [inline actions/tahoe-lafs]: "if needed"

File: actions/tahoe-lafs [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+    if not os.path.exists(domain_name_file):
+        with open(domain_name_file, 'w') as dnf:
+            dnf.write(arguments.domain_name)
+
+
+def subcommand_autostart(_):
+    """Automatically start all introducers and storage nodes on system startup.
+    """
+    with open('/etc/default/tahoe-lafs', 'a') as autostart_conf:
+        autostart_conf.write('AUTOSTART="all"')
```
Comment (jvalleroy) [inline actions/tahoe-lafs]: We should make this idempotent, by first checking if the file contains this line already.

File: data/etc/apache2/conf-available/tahoe-plinth.conf [hunk 9]
```diff
@@ -0,0 +1,11 @@
+# Tahoe-LAFS Storage Node web interface
+
+# See if Tahoe-LAFS has basic auth
+# Enable Tahoe only from localhost
+# Do basic auth and LDAP from apache reverse proxy
```
Comment (jvalleroy) [inline data/etc/apache2/conf-available/tahoe-plinth.conf]: Are these TODOs? I think we will want to have login required at least (through apache)

File: plinth/actions.py [hunk 10]
```diff
@@ -118,7 +118,14 @@ def superuser_run(action, options=None, input=None, async=False):
...
 
-def _run(action, options=None, input=None, async=False, run_as_root=False):
+def run_as_user(action, options=None, input=None, async=False, become_user=None):
+    """
+    Run a command as a different user. If become_user is None, run as current user.
+    """
+    return _run(action, options, input, async, False, become_user)
+
+
+def _run(action, options=None, input=None, async=False, run_as_root=False, become_user=None):
```
Comment (jvalleroy) [inline plinth/actions.py]: Split this long line, just for code style.

File: plinth/actions.py [hunk 11]
```diff
@@ -159,6 +166,8 @@ def _run(action, options=None, input=None, async=False, run_as_root=False):
     # Contract 1: commands can run via sudo.
     if run_as_root:
         cmd = ['sudo', '-n'] + cmd
+    elif become_user:
+        cmd = ['sudo', '-u', become_user] + cmd
```
Comment (jvalleroy) [inline plinth/actions.py]: I think we should keep the '-n' here also.

File: plinth/domain_selection_form.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+
+from plinth import utils
+from django import forms
+from django.utils.translation import ugettext_lazy as _
+
+
+class DomainSelectionForm(forms.Form):
```
Comment (jvalleroy) [inline plinth/domain_selection_form.py]: I think this can be moved into plinth/forms.py.

File: plinth/exceptions.py [hunk 13]
```diff
@@ -0,0 +1,2 @@
+class DomainNameNotSetupException(Exception):
```
Comment (jvalleroy) [inline plinth/exceptions.py]: This can be moved into plinth/errors.py.

File: plinth/modules/tahoe/__init__.py [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+    introducer = actions.run_as_user('tahoe-lafs', ['get-local-introducer'],
+                                     become_user='tahoe-lafs')
+
+    return json.loads(introducer)
+
+
+def restart_storage_node():
+    try:
+        os.chdir(tahoe_home)
+        subprocess.check_output(
```
Comment (jvalleroy) [inline plinth/modules/tahoe/__init__.py]: Should use actions here.

File: plinth/modules/tahoe/templates/tahoe-post-setup.html [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+    {% blocktrans trimmed with domain_name=domain_name %}
+    The Tahoe-LAFS server domain is set to <b>{{ domain_name }}</b>.
+    Changing the FreedomBox domain name needs a reinstall of
+    Tahoe-LAFS and you WILL LOSE DATA.
+    {% endblocktrans %}
+</p>
+{% endblock %}
+
+{% block configuration %}
+  <h3>Configuration</h3>
```
Comment (jvalleroy) [inline plinth/modules/tahoe/templates/tahoe-post-setup.html]: We will need to internationalize this string (and others in this file).

File: plinth/modules/tahoe/templates/tahoe-pre-setup.html [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+
+  {% block status %}
+  {% endblock %}
+
+  {% block diagnostics %}
+  {% endblock %}
+
+  {% block configuration %}
+    {% if domain_names|length > 0 %}
+      <h3>Configuration</h3>
```
Comment (jvalleroy) [inline plinth/modules/tahoe/templates/tahoe-pre-setup.html]: We will need to internationalize this string (and others in this file).

File: plinth/utils.py [hunk 17]
```diff
@@ -63,3 +66,49 @@ def is_user_admin(request, cached=False):
     user_is_admin = request.user.groups.filter(name='admin').exists()
     request.session['cache_user_is_admin'] = user_is_admin
     return user_is_admin
+
+
+def get_domain_names():
```
Comment (jvalleroy) [inline plinth/utils.py]: Perhaps we can replace the extra get_domain_names functions in matrix-synapse and diaspora modules.

**Comment by lispyclouds** [comment]
@jvalleroy Implemented.

**Comment by jvalleroy** [comment]
@JosephKiranBabu @lispyclouds I made some updates and pushed to https://github.com/jvalleroy/Plinth/tree/tahoe-lafs.

There are a few remaining issues:
- The logo appears distorted on the front page.
- I couldn't use HTTPS to access Tahoe-LAFS. Does it need to be added to the apache config?
- Would it make sense to use a subdomain, instead of another port? Similar to diaspora module.

**Comment by jvalleroy** [comment]
I merged this pull requests (after squashing commits). I'll open new issues to track the remaining TODOs.

**Pull request closed without merging**